pallets/xcm-asset-manager/src/lib.rs

```diff
@@ -0,0 +1,457 @@
+	/// This is mostly used when receiving transaction specifying an asset directly,
+	/// like transferring an asset from this chain to another.
+	#[pallet::storage]
+	#[pallet::getter(fn asset_id_type)]
+	pub type AssetIdType<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, T::ForeignAssetType>;
+
+	/// Reverse mapping of AssetIdType. Mapping from an asset type to an asset id.
+	/// This is mostly used when receiving a multilocation XCM message to retrieve
+	/// the corresponding asset in which tokens should me minted.
+	#[pallet::storage]
+	#[pallet::getter(fn asset_type_id)]
+	pub type AssetTypeId<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::ForeignAssetType, T::AssetId>;
+
+	/// Stores the units per second for local execution for a AssetType.
+	/// This is used to know how to charge for XCM execution in a particular asset
+	/// Not all assets might contain units per second, hence the different storage
+	#[pallet::storage]
+	#[pallet::getter(fn asset_id_units_per_second)]
+	pub type AssetIdUnitsPerSecond<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, u128>;
+
+	/// Stores the tracker of foreign assets id that have been
+	/// created so far
+	#[pallet::storage]
+	#[pallet::getter(fn foreign_asset_tracker)]
+	pub type ForeignAssetTracker<T: Config> = StorageValue<_, T::AssetId, ValueQuery>;
+
+	// Supported fee asset payments
+	#[pallet::storage]
+	#[pallet::getter(fn supported_fee_payment_assets)]
+	pub type SupportedFeePaymentAssets<T: Config> = StorageValue<_, Vec<T::AssetId>, ValueQuery>;
+
+	/// The storages for AssetIdMetadatas.
+	///
+	/// AssetIdMetadatas: map AssetId => Option<AssetMetadata>
+	#[pallet::storage]
+	#[pallet::getter(fn asset_metadatas)]
+	pub type AssetIdMetadatas<T: Config> =
+		StorageMap<_, Twox64Concat, T::AssetId, AssetMetadata<BalanceOf<T>>, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register new asset with the asset manager
+		/// TODO::Reserve native token multilocation through GenesisBuild/RuntimeUpgrade
+		#[pallet::weight(T::WeightInfo::register_foreign_asset())]
+		pub fn register_foreign_asset(
+			origin: OriginFor<T>,
+			asset_type: T::ForeignAssetType,
+			metadata: AssetMetadata<BalanceOf<T>>,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			let mut asset_id = ForeignAssetTracker::<T>::get();
+
+			ensure!(!AssetTypeId::<T>::contains_key(&asset_type), Error::<T>::AssetAlreadyExists);
+
+			AssetIdMetadatas::<T>::insert(&asset_id, &metadata);
+			AssetIdType::<T>::insert(&asset_id, &asset_type);
+			AssetTypeId::<T>::insert(&asset_type, &asset_id);
+
+			Self::deposit_event(Event::<T>::ForeignAssetRegistered { asset_id, asset_type });
+			Self::deposit_event(Event::<T>::ForeignAssetMetadataUpdated { asset_id, metadata });
+
+			asset_id = asset_id.checked_add(&One::one()).ok_or(Error::<T>::AssetIdOverflow)?;
+			// Auto increment for Asset counter
+			ForeignAssetTracker::<T>::put(asset_id);
```

Comment:
I am not sure if I understand it correctly.

So do you mean asset_id is consumed by deposit_event, so can not be later used?
I think it will not, see the "collator_count" usage of here.
https://github.com/PureStake/moonbeam/blob/master/pallets/parachain-staking/src/lib.rs#L429
I thought use a variable to combine a event struct and use the struct to a function (deposit_event) will not consume the variable.

Reply:
Thanks!